content/program-security.md

```diff
@@ -0,0 +1,741 @@
+
+```rust
+if !account_data.is_initialized() {
+    msg!("Account is not initialized");
+    return Err(ProgramError::AccountAlreadyInitialized);
+}
+```
+
+But wait, take a closer look at the `ProgramError` we’re returning there - it says `AccountAlreadyInitialized` which made sense in the `add_movie_review` function but doesn’t really make sense here since the account *should* be initialized already. Here is an example of when you might want to create your own custom program errors and we’re going to do just that now.
+
+### 3. Custom Program Errors
+
+To account for this and make our errors a little more relevant, create a new file called error.rs and add the following custom errors.
+
+```rust
+use solana_program::{program_error::ProgramError};
+use thiserror::Error;
+
+/// Errors that may be returned by the MovieReview program.
+#[derive(Debug, Error)]
+pub enum ReviewError{
+    // 0
+    /// UnitializedAccount
+    #[error("Account not initialized yet")]
+    UnitializedAccount,
+    // 1
+    // Incorrect PDA
+    #[error("PDA derived does not equal PDA passed in")]
+    InvalidPDA,
+    // 2
+    // Data is too long
+    #[error("Input data exceeds max length")]
+    InvalidDataLength,
+    //3
+    // Rating is too high
+    #[error("Rating cannot exceed 5")]
+    RatingTooHigh,
+}
+
+impl From<ReviewError> for ProgramError {
+    fn from(e: ReviewError) -> Self {
+        ProgramError::Custom(e as u32)
+    }
+}
+```
+
+Then bring in the following crate to processor.rs like so
+
+```rust
+// inside processor.rs
+use crate::error::ReviewError;
+```
+
+### 4.Account Validation Checks
+
+Now you can refactor the `is_initialized` validation check to return the new `UnitializedAccount` error type that we just created.
+
+```rust
+msg!("checking if movie account is initialized");
+if !account_data.is_initialized() {
+  msg!("Account is not initialized");
+  return Err(ReviewError::UnitializedAccount.into());
+}
+```
+
+Our next validation check will be on the pda itself. We will derive the pda ourselves with the initializer’s public key and the title that is currently stored in the movie review. The reason we’re doing this is because you should only be able to update a review that was created by you and since the way the pdas were derived in the first place is with the initializer’s key and the title, this is an implicit check that the `initializer` passed in here did indeed create this review. If the `initializer` provided here is not the same public key that was used to derive this pda, then the pda that we derive program side will not equal the pda that was passed in. We’ll also return the second custom error type that we defined in [error.rs](http://error.rs) if they don’t match.
```

Comment:
@jamesrp13 do we want to use uppercase acronyms for PDA?

Project-wide, outside of this lesson, we have no uses of lowercase "pda" within a (non code) sentence, and ~20 uses capitalized.  Lemme know and I'll sort them out when yr wrapping up